src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,325 @@
+    BaseFloat samp_freq_input = input_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_input = input_matrix.NumCols(),  // number of samples in recording
+          num_input_channel = input_matrix.NumRows();  // number of channels in recording
+    KALDI_VLOG(1) << "sampling frequency of input: " << samp_freq_input
+                  << " #samples: " << num_samp_input
+                  << " #channel: " << num_input_channel;
+    KALDI_ASSERT(input_channel < num_input_channel);
+
+    WaveData rir_wave;
+    {
+      Input ki(rir_file);
+      rir_wave.Read(ki.Stream());
+    }
+    const Matrix<BaseFloat> &rir_matrix = rir_wave.Data();
+    BaseFloat samp_freq_rir = rir_wave.SampFreq();
+    int32 num_samp_rir = rir_matrix.NumCols(),
+          num_rir_channel = rir_matrix.NumRows();
+    KALDI_VLOG(1) << "sampling frequency of rir: " << samp_freq_rir
+                  << " #samples: " << num_samp_rir
+                  << " #channel: " << num_rir_channel;
+    if (!multi_channel_output) {
+      KALDI_ASSERT(rir_channel < num_rir_channel);
+    }
+
+    Matrix<BaseFloat> noise_matrix;
+    if (!noise_file.empty()) {
+      WaveData noise_wave;
+      {
+        Input ki(noise_file);
+        noise_wave.Read(ki.Stream());
+      }
+      noise_matrix = noise_wave.Data();
+      BaseFloat samp_freq_noise = noise_wave.SampFreq();
+      int32 num_samp_noise = noise_matrix.NumCols(),
+            num_noise_channel = noise_matrix.NumRows();
+      KALDI_VLOG(1) << "sampling frequency of noise: " << samp_freq_noise
+                    << " #samples: " << num_samp_noise
+                    << " #channel: " << num_noise_channel;
+      if (multi_channel_output) {
+        KALDI_ASSERT(num_rir_channel == num_noise_channel);
+      } else {
+        KALDI_ASSERT(noise_channel < num_noise_channel);
+      }
+    }
+
+    int32 num_output_channel = (multi_channel_output ? num_rir_channel : 1);
+    Matrix<BaseFloat> out_matrix(num_output_channel, num_samp_input);
+
+    for (int32 output_channel = 0; output_channel < num_output_channel; output_channel++) {
+      Vector<BaseFloat> input(num_samp_input);
+      input.CopyRowFromMat(input_matrix, input_channel);
+
+      int32 this_rir_channel = (multi_channel_output ? output_channel : rir_channel);
+      Vector<BaseFloat> rir(num_samp_rir);
+      rir.CopyRowFromMat(rir_matrix, this_rir_channel);
+      rir.Scale(1 / MaxAbsolute(rir));
+
+      Vector<BaseFloat> noise(0);
+      if (!noise_file.empty()) {
+        noise.Resize(noise_matrix.NumCols());
+        int32 this_noise_channel = (multi_channel_output ? output_channel : noise_channel);
+        noise.CopyRowFromMat(noise_matrix, this_noise_channel);
+      }
+
+      DoReverberation(rir, &noise, &input, samp_freq_rir, snr_db, test_mode);
+
+      float input_max = MaxAbsolute(input);
+      for (int32 j = 0; j < input.Dim(); j++)
+        input(j) = (0.99 * input(j) / input_max) * pow(2, 15);
```

Comment:
BTW, another amusing comment on the WAV/RIFF specification (which might
explain some of the problems it has caused us): from
https://en.wikipedia.org/wiki/WAV,
"The RIFF specification attempts to be a formal specification, but its
formalism lacks the precision seen in other tagged formats."
I suppose this is an instance of the hacker mentality and disdain for
formal education that Microsoft had in its early years.




On Fri, Aug 21, 2015 at 10:28 PM, Kirill Katsnelson <
notifications@github.com> wrote:

On Sat, Aug 22, 2015 at 11:16 AM, Daniel Povey <dpovey@gmail.com> wrote:

> Kirill, I don't think there's any point doing the 8-bit training.
> Probably no-one ever successfully used the 8-bit reading code.
> I could not find any official Microsoft documentation of the fact that the
> 8-bit PCMs are unsigned, but according to
> https://en.wikipedia.org/wiki/WAV,
> "There are some inconsistencies in the WAV format: for example, 8-bit
> data is unsigned while 16-bit data is signed, and many chunks duplicate
> information found in other chunks."
> We already support only a subset of the WAV format, and I'd be happier to
> go to an even smaller subset and require people to format their data to
> 16-bit linear PCM using sox or similar.  Let's just remove the 8-bit
> support, because it looks like it never would have worked.  Also according
> to http://wiki.multimedia.cx/?title=PCM
> it looks like 32-bit PCM is very rare.  So I think we should just remove
> the 32-bit support too, and declare that the WaveReader supports only
> 16-bit PCM and does not do any rescaling.
> @vpeddinti, please note!
> Dan
>
>
>
> In src/featbin/wav-reverberate.cc
>> <https://github.com/kaldi-asr/kaldi/pull/76#discussion_r37692856>:
>>
>> > +      Vector<BaseFloat> rir(num_samp_rir);
>> > +      rir.CopyRowFromMat(rir_matrix, this_rir_channel);
>> > +      rir.Scale(1 / MaxAbsolute(rir));
>> > +
>> > +      Vector<BaseFloat> noise(0);
>> > +      if (!noise_file.empty()) {
>> > +        noise.Resize(noise_matrix.NumCols());
>> > +        int32 this_noise_channel = (multi_channel_output ? output_channel : noise_channel);
>> > +        noise.CopyRowFromMat(noise_matrix, this_noise_channel);
>> > +      }
>> > +
>> > +      DoReverberation(rir, &noise, &input, samp_freq_rir, snr_db, test_mode);
>> > +
>> > +      float input_max = MaxAbsolute(input);
>> > +      for (int32 j = 0; j < input.Dim(); j++)
>> > +        input(j) = (0.99 * input(j) / input_max) * pow(2, 15);
>>
>> I'll do the fixing then, if it is not very urgent. Seems that this PR
>> will not be affected.
>>
>> The term is "clip," not "truncate." Just FYI so it won't have
>> inadvertently snuck into function names or comments.
>>
>> What I want to do is try training a system on 8 bit samples (before
>> fixing the code). Naively it seems that this should be a total wreck, but
>> sometimes ML can extract information from an impossible mess (the incorrect
>> sample decoding as it currently is does not lose information per se, as,
>> amazingly, does not FFT). I just want to see what happens!
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/76/files#r37692856>.
>>
>
>